src/main/java/org/prebid/server/handler/SetuidHandler.java

```diff
@@ -40,15 +50,63 @@ public void handle(RoutingContext context) {
         final String bidder = context.request().getParam("bidder");
         if (StringUtils.isBlank(bidder)) {
             final int status = HttpResponseStatus.BAD_REQUEST.code();
-            context.response().setStatusCode(status).end();
+            context.response().setStatusCode(status).end("\"bidder\" query param is required");
             metrics.cookieSync().incCounter(MetricName.bad_requests);
             analyticsReporter.processEvent(SetuidEvent.error(status));
             return;
         }
 
+        final String gdpr = context.request().getParam("gdpr");
+        final String gdprConsent = context.request().getParam("gdpr_consent");
+        final String ip = useGeoLocation ? HttpUtil.ipFrom(context.request()) : null;
+        gdprService.analyze(gdpr, gdprConsent, ip)
+                .setHandler(result -> handleResult(result, context, uidsCookie, bidder, gdpr, gdprConsent));
+    }
+
+    private void handleResult(AsyncResult<GdprResponse> result, RoutingContext context, UidsCookie uidsCookie,
+                              String bidder, String gdpr, String gdprConsent) {
+        final GdprResult gdprResult = result.result().getGdprResult();
+
+        if (gdprResult == GdprResult.allowed) {
+            respondWithCookie(context, bidder, uidsCookie);
+        } else {
+            respondWithoutCookie(gdprResult, context, bidder, gdpr, gdprConsent);
+        }
+    }
+
+    private void respondWithoutCookie(GdprResult gdprResult, RoutingContext context, String bidder, String gdpr,
+                                      String gdprConsent) {
+        final int status;
+        final String body;
+
+        switch (gdprResult) {
+            case error_invalid_gdpr:
```

Comment:
Using enum to specify error type is a little bit alien for java. Is it possible to replace them with exceptions?